### PR TITLE
Add SAS token generator service

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-869

### Change description ###
- **Add SAS Token generator service**
- In the next PRs, 
   - Controller for SAS token endpoint
   - Config changes
   - Functional tests

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -12,3 +12,8 @@ management:
 spring:
   application:
     name: Blob Router Service
+
+accesstoken:
+  serviceConfig:
+    - serviceName: bulkscan
```
Comment (mdayican) [inline src/main/resources/application.yaml]: came case  or kebab-case?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        SharedAccessBlobPolicy policy = new SharedAccessBlobPolicy();
+        policy.setPermissions(EnumSet.of(WRITE, LIST));
+        policy.setSharedAccessExpiryTime(Date.from(now().plusSeconds(config.getValidity()).toInstant(ZoneOffset.UTC)));
+
+        return policy;
+    }
+
+    private TokenConfig getTokenConfigForService(String serviceName) {
+        return accessTokenProperties.getServiceConfig().stream()
```
Comment (mdayican) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java]: do we need to search every time? we can keep it as a map?

File: src/main/resources/application.yaml (hunk 1 above)
Comment (doncem) [reply src/main/resources/application.yaml]: those will be all lowercase - same as we have currently in processor. please check bulk-scan-shared-infra for references ;)

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    @Test
+    void should_generate_sas_token_when_service_configuration_is_available() {
+        String sasToken = tokenGeneratorService.generateSasToken("bulkscan");
+
+        String decodedSasToken = Utility.urlDecode(sasToken);
+        Map<String, String[]> queryParams = StorageImplUtils.parseQueryStringSplitValues(decodedSasToken);
+        String currentDate = new SimpleDateFormat("yyyy-MM-dd").format(new Date());
+
+        assertThat(queryParams.get("sig")).isNotNull();//this is a generated hash of the resource string
+        assertThat(queryParams.get("se")[0]).startsWith(currentDate);//the expiry date/time for the signature
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java]: can we check time as well, to make sure that `300` configured is setup is actually used/relevant?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    private Map<String, ServiceConfig> servicesConfig;
+
+    public Map<String, ServiceConfig> getServicesConfig() {
+        return servicesConfig;
+    }
+
+    public void setServicesConfig(List<ServiceConfig> servicesConfig) {
+        this.servicesConfig = servicesConfig
+            .stream()
+            .collect(Collectors.toMap(ServiceConfig::getName, config -> config));
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java]: Let's make it smarter and adapt functional interfaces as per recommendation. So first bit is already using the reference, second one is screaming for `Function::identity` or whichever the signature is. Of course - giving stream collector allows

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    public SasTokenGeneratorService(
+        StorageSharedKeyCredential storageSharedKeyCredential,
+        ServiceConfiguration serviceConfiguration
+    ) {
+        this.storageSharedKeyCredential = storageSharedKeyCredential;
+        this.serviceConfiguration = serviceConfiguration;
+    }
+
+    public String generateSasToken(String serviceName) {
+        log.info("SAS Token request received for service {} ", serviceName);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java]: Entry to service method theoretically can originate from anywhere. Perhaps we can just log the fact that we are `Generating SAS Token for {} service`?

File: src/main/resources/application.yaml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 server:
   port: 8584
 
-management:
+  management:
```
Comment (doncem) [inline src/main/resources/application.yaml]: accident?

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -12,3 +12,7 @@ management:
 spring:
   application:
     name: Blob Router Service
+
+service:
+  - name: bulkscan
+    sas-validity: ${SAS_TOKEN_VALIDITY:300} #In seconds
```
Comment (doncem) [inline src/main/resources/application.yaml]: In processor we actually never changed this value. Is there a point for us introducing extra flexibility? Let's keep it simple and less headache configuring charts, if we need in future - we can introduce customisations

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import uk.gov.hmcts.reform.blobrouter.exceptions.ServiceConfigNotFoundException;
+
+import java.time.OffsetDateTime;
+import java.util.Map;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (doncem) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java]: There's nothing mocked here, perhaps we don't need extension at all

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            "testAccountName", "dGVzdGtleQ=="
+        );
+
+        createAccessTokenConfig();
+
+        tokenGeneratorService = new SasTokenGeneratorService(
+            storageCredentials,
+            serviceConfiguration
+        );
+    }
```
Comment (doncem) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java]: Is it possible/beneficial to move to `@BeforeAll` section - load everything statically? `storageCredentials` is automatically top level private static field. the rest can be set up once per test - also statically

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+        BlobServiceSasSignatureValues sasSignatureBuilder = getBlobServiceSasSignatureValues(serviceName);
+        return sasSignatureBuilder.generateSasQueryParameters(storageSharedKeyCredential).encode();
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java]: Do we want to wrap return statement with try/catch here? And actually use currently not used `UnableToGenerateSasTokenException`?

File: src/main/resources/application.yaml (hunk 1 above)
Comment (doncem) [reply src/main/resources/application.yaml]: oh sorry, i think i confused myself. clarified with @aliveni's help

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java (hunk 3 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java]: updated test to verify the expiry time

File: build.gradle [hunk 11]
```diff
@@ -189,7 +189,8 @@ ext.libraries = [
 ]
 
 dependencies {
-  compile group: 'com.microsoft.azure', name: 'azure-storage', version: '8.4.0'
+  compile group: 'com.microsoft.rest.v2', name: 'client-runtime', version: '2.1.1'
```
[the remark this answers is not in the document]
Comment (aliveni) [reply build.gradle]: Removed the library

File: build.gradle [hunk 12]
```diff
@@ -206,6 +209,8 @@ dependencies {
     exclude group: 'junit', module: 'junit'
     exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
   }
+  testCompile group: 'org.mockito', name: 'mockito-junit-jupiter', version: '3.1.0'
+  testCompile group: 'org.assertj', name: 'assertj-core', version: '3.14.0'
```
[the remark this answers is not in the document]
Comment (aliveni) [reply build.gradle]: removed both assertj and mockito libraries

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java (hunk 5 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java]: Updated the message

File: src/main/resources/application.yaml (hunk 6 above)
Comment (aliveni) [reply src/main/resources/application.yaml]: yeah, alignment is fixed now

File: src/main/resources/application.yaml (hunk 7 above)
Comment (aliveni) [reply src/main/resources/application.yaml]: ok, I've set to 300 seconds for now (Same as  processor)

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java (hunk 8 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java]: removed

File: src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java (hunk 9 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorServiceTest.java]: Done

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java (hunk 10 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java]: Added  the exception and a test for this scenario

File: src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java (hunk 4 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/config/ServiceConfiguration.java]: Done

File: src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        return new BlobServiceSasSignatureValues()
+            .setProtocol(SasProtocol.HTTPS_ONLY) // Users MUST use HTTPS (not HTTP).
+            .setContainerName(serviceName)
+            .setExpiryTime(OffsetDateTime.now().plusSeconds(config.getSasValidity()))
+            .setPermissions(permissions);
+    }
+
+    private ServiceConfig getConfigForService(String serviceName) {
+        if (!serviceConfiguration.getServicesConfig().containsKey(serviceName)) {
+            throw new ServiceConfigNotFoundException("No service configuration found for service " + serviceName);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/services/SasTokenGeneratorService.java]:
```suggestion
            throw new ServiceConfigNotFoundException("No service configuration found for " + serviceName);
```

**Pull request merged**